### PR TITLE
Add user group association

### PR DESCRIPTION
**Pull request by thomasiles**
# Users can belong to Groups

This is part of the user management work adding the concept of groups to connect users, organisations and forms.
 
Trello card: https://trello.com/c/elvPhWEc/1381-add-association-between-groups-users-and-forms-slice-2
- Users can belong to zero or more groups

This PR adds a new model and DB table, `Membership`. This table joins 'User' to `Group`.

There are some business rules we want to enforce with this relationship:
- Users can belong to zero or more groups
- Users can only belong to a single group once
- Users can only belong to a Group if they are in the same Organisation as the Group

The first two are handled by unique constraints in the ruby models and at in the database, through the schema design and indexes.

The third cannot easily be enforced at the database level using our current schema. It's enforced only through the validation on the Membership class.

## Users can only belong to a Group if they are in the same Organisation as the Group

We use this validation on Membership to make sure user and organisation match when changing the membership.

```ruby
  def user_and_group_in_same_organisation
    return if user.nil? || group.nil?
    return if user.organisation == group.organisation

    errors.add :base, "User and group must have the same organisation"
  end
``` 

To enforce this, I think the we would need to add `organisation_id` to `Membership` and use it in composite foreign keys on `User` and `Group`. This would prevent users who are not part of the same organisation as the group being referenced.

### What happens without this?
If a `User` joins a `Group` but later changes their organisation, they will still be part of the group in the DB.

### How do we mitigate it?
I think we have to deal with this in the user and group model to update membership when organisation is changed.  This PR doesn't tackle it yet.

### Why not change the schema to include organisation?

It's a bit of a heavyweight change and commits us to a multi-tenant strategy that I'm not sure we are ready for. It's not completely clear how central organisation will end up being at this moment in the design/development so I think we should hold off.

### Things to consider when reviewing

- Ensure that you consider the wider context.
- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Are there all the unit tests needed?
- Has all relevant documentation been updated?

### CODE REVIEW

File: app/models/group.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 class Group < ApplicationRecord
   belongs_to :organisation
 
+  has_many :memberships, dependent: :destroy
```
Comment (thomasiles) [inline app/models/group.rb:4]: if a group is deleted, so are its memberships

File: app/models/membership.rb [hunk 2]
```diff
@@ -2,4 +2,15 @@ class Membership < ApplicationRecord
   belongs_to :user
   belongs_to :group
   belongs_to :creator, class_name: "User"
+
+  validate :user_and_group_in_same_organisation
```
Comment (lfdebrux) [inline app/models/membership.rb:6]: Can we update the commit message for 80f4cd86a38b0df95b6009a36a04bf8a93dca8bc? It says 'Prevent a user being a member of a group more than once.', but the changes appear to be about a different validation.

**Review by lfdebrux — COMMENTED** [review]
I've read through the code, looks good at first glance, I want to do some testing though...

File: db/migrate/20240223134039_create_memberships.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class CreateMemberships < ActiveRecord::Migration[7.1]
+  def change
+    create_table :memberships do |t|
+      t.references :user, null: false, foreign_key: true, comment: "The user who is a member of the group"
+      t.references :group, null: false, foreign_key: true, comment: "The group that the user is a member of"
+      t.references :creator, null: false, foreign_key: { to_table: :users }, comment: "The user who created the membership"
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply db/migrate/20240223134039_create_memberships.rb:6]: Sorry to bikeshed, thinking about it further, `creator` feels weird to me, what about `added_by` as a name for this attribute? It feels more natural to say one user added another, and I've seen that particular term used in other Rails codebases [[1](https://github.com/bullet-train-co/bullet_train-core/blob/9227d144096d0990d5abb4a90be994c2faeeb77f/bullet_train/db/migrate/20200219013834_add_added_by_to_membership.rb#L3)].

File: db/migrate/20240223134039_create_memberships.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class CreateMemberships < ActiveRecord::Migration[7.1]
+  def change
+    create_table :memberships do |t|
+      t.references :user, null: false, foreign_key: true, comment: "The user who is a member of the group"
+      t.references :group, null: false, foreign_key: true, comment: "The group that the user is a member of"
```
[the remark this answers is not in the document]
Comment (thomasiles) [reply db/migrate/20240223134039_create_memberships.rb:5]: I think we probably can get rid of them 👍

**Comment by thomasiles** [comment]
Thanks for reviewing, I've updated it with the suggestions above 👍

File: app/models/membership.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  belongs_to :user
+  belongs_to :group
+  belongs_to :added_by, class_name: "User"
+
+  validate :user_and_group_in_same_organisation
+  validates :user, uniqueness: { scope: :group }
+
+private
+
+  def user_and_group_in_same_organisation
```
Comment (chao-xian) [inline app/models/membership.rb:11]: The commit message says the code was from the Rails generator, but as this is custom it should probably go into a separate commit with the appropriate spec tests.

File: spec/factories/models/memberships.rb
Comment (chao-xian) [inline spec/factories/models/memberships.rb:1]: The commit title says `...and move specs` but I wouldn't call this a spec...?

**Review by chao-xian — COMMENTED** [review]
Just some commit tidy ups from me but looking good so far.

**Comment by lfdebrux** [comment]
@thomasiles thanks for writing up the bits about organisation membership in more detail, that's very helpful.

Thinking about what to do if a user's organisation changes, it might be a bit drastic, but what if we just say for know that when a user moves from one organisation to another, they lose all of their group memberships? I think that would make some of the questions you had moot, what do you think?

It should be straightforward to add a step to the `on_user_update` method to achieve this:

https://github.com/alphagov/forms-admin/blob/7d6d76d28ff9edefaf23c38cd2fa3228debb3719/app/service/user_update_service.rb#L15-L18

**Comment by thomasiles** [comment]
> @thomasiles thanks for writing up the bits about organisation membership in more detail, that's very helpful.
> 
> Thinking about what to do if a user's organisation changes, it might be a bit drastic, but what if we just say for know that when a user moves from one organisation to another, they lose all of their group memberships? I think that would make some of the questions you had moot, what do you think?
> 
> It should be straightforward to add a step to the `on_user_update` method to achieve this:
> 
> https://github.com/alphagov/forms-admin/blob/7d6d76d28ff9edefaf23c38cd2fa3228debb3719/app/service/user_update_service.rb#L15-L18

yes - i think this is the only thing which makes sense. I'd rather not make this commit any longer though so I think it's best to do it in another one.

**Review by lfdebrux — APPROVED** [review]
